microhapulator/api.py

```diff
@@ -739,3 +739,57 @@ def off_target_mapping(marker_bam_file, fullref_bam_file, markertsv, minbasequal
         counts["OffTargetReads"].append(off_target_count)
     data = pd.DataFrame(counts).sort_values(by="Marker").reset_index(drop=True)
     return data
+
+
+def donut_plot(marker_mapped, refr_mapped, off_target_mapped, output, sample=None):
+    """Make donut plot of proportions on target, off target, repetitive, and contaminant reads
+
+    :param str marker_mapped: path of txt file containing number of reads mapped to marker sequences
+    :param str refr_mapped: path of txt file containing number of reads mapped to the full human reference genome
+    :param str off_target_mapped: path of txt file containing number of reads mapped preferentially to non-marker loci in the human reference genome
+    :param str output: path where the png file of the plot will be saved
+    :param str sample: name of the sample to be included as the plot title; by default no sample name is shown
+    """
+    data = count_mapped_read_types(marker_mapped, refr_mapped, off_target_mapped)
+    percs = data / sum(data) * 100
+    backend = matplotlib.get_backend()
+    plt.switch_backend("Agg")
+    plt.figure(figsize=(10, 4), dpi=300)
+    labels = ["on target", "repetitive", "off target", "contamination"]
+    wedges, texts = plt.pie(percs, startangle=1)
+    kw = dict(arrowprops=dict(arrowstyle="-", linewidth=0.5), zorder=0, va="center")
+    for i, p in enumerate(wedges):
+        ang = (p.theta2 - p.theta1) / 2.0 + p.theta1
+        y = np.sin(np.deg2rad(ang))
+        x = np.cos(np.deg2rad(ang))
+        horizontalalignment = {-1: "right", 1: "left"}[int(np.sign(x))]
+        connectionstyle = "angle,angleA=0,angleB={}".format(ang)
+        kw["arrowprops"].update({"connectionstyle": connectionstyle})
+        plt.annotate(
+            f'{labels[i]} ({"{:.2f}".format(percs[i])}%)',
+            xy=(x, y),
+            xytext=(1.4 * np.sign(x), 1.4 * y),
+            horizontalalignment=horizontalalignment,
+            **kw,
+        )
```

Comment:
Adds annotations to each "wedge" of the donut plot with the label and the percentage of reads comprising that wedge.

Reply:
These annotations are a nice improvement over the donut plots I implemented for ReadQuant. Kudos for the geometric calculations here! :smiley:

Reply:
Kudos to someone on Stack Overflow for those calculations 😉

Reply:
If you can find that thread again, I'd recommend dropping a link in the code. Example from MicroHapulator.

https://github.com/bioforensics/MicroHapulator/blob/8bee2d168426b5acf1083ae8c9d5fa83742c4706/microhapulator/cli/getrefr.py#L23-L29

Exact wording is up to you. I'll often do "Adapted from URL" or "Stolen shamelessly from URL."